### PR TITLE
Eo3 only index drivers

### PR DESCRIPTION
**Pull request by SpacemanPaul**
### Reason for this pull request

Progressing on the ODCv2 roadmap,  refactors in preparation for making Postgis index driver EO3 only.

### Proposed changes

- Refactor and cleanup of `index/eo3.py`
- Metadata class format support flags in index driver:
  - `supports_persistance` (i.e. add/update/archive/purge work to some degree)
  - `supports_legacy` (i.e. non-eo3 geospatial)
  - `supports_nongeo` (e.g. telemetry)
  - Note that eo3 support is assumed. There is no explicit `supports_eo3` flag.
- Improve type hints in `index/eo3.py` and `index/hl.py`

This means that a lot of existing parametized tests won't work for the postgis index driver any more, as they rely on non-EO3 legacy test-data.  Despite this coverage is mostly unaffected.  Note that large drop in coverage on `postgis/_datasets.py` is due to removing well-tested code rather than adding poorly tested code.

 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

File: datacube/index/abstract.py [hunk 1]
```diff
@@ -909,30 +909,38 @@ class AbstractIndex(ABC):
...
+    #   supports legacy ODCv1 EO style metadata types.
+    supports_legacy = True
+    #   supports non-geospatial (e.g. telemetry) metadata types
+    supports_nongeo = True
+
     @property
     @abstractmethod
     def url(self) -> str:
-        pass
+        ...
```
Comment (omad) [inline datacube/index/abstract.py:923]: My preferred style here is to include a docstring. Although, for a property it's of dubious utility.

File: datacube/utils/changes.py [hunk 2]
```diff
@@ -93,7 +93,7 @@ def get_doc_changes(original: Changable,
         for idx, (orig_item, new_item) in enumerate(zip_longest(original, new)):
             changed_fields.extend(get_doc_changes(orig_item, new_item, base_prefix + (idx, )))
     elif isinstance(original, tuple) or isinstance(new, tuple):
-        if not numpy.array_equal(original, new):
+        if not numpy.array_equal(cast(Sequence[Any], original), cast(Sequence[Any], new)):
```
Comment (omad) [inline datacube/utils/changes.py:96]: :exploding_head:

**Review by omad — APPROVED** [review]
I think this looks good. Thanks Paul

File: datacube/index/abstract.py (hunk 1 above)
Comment (SpacemanPaul) [reply datacube/index/abstract.py:923]: Good point, `AbstractIndex` and `AbstractIndexDriver` could both do with some inline documentation.

**Pull request merged**